src/Altinn.App.Core/AppImplementationFactory.cs

```diff
@@ -0,0 +1,57 @@
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Altinn.App.Core.Features;
+
+/// <summary>
+/// Marker attribute for interfaces that are meant to be implemented by apps.
+/// </summary>
+[AttributeUsage(AttributeTargets.Interface, AllowMultiple = false)]
+internal sealed class ImplementableByAppsAttribute : Attribute { }
+
+internal static class AppImplementationFactoryExtensions
+{
+    public static IServiceCollection AddAppImplementationFactory(this IServiceCollection services)
+    {
+        services.AddSingleton<AppImplementationFactory.DIScopeHolder>(sp =>
+            new(sp.GetRequiredService<IHttpContextAccessor>().HttpContext?.RequestServices)
+        );
+        services.AddSingleton<AppImplementationFactory>();
+        return services;
+    }
+
+    public static IServiceCollection AddTestAppImplementationFactory(this IServiceCollection services)
+    {
+        services.AddSingleton<AppImplementationFactory.DIScopeHolder>(sp => new(sp));
+        services.AddSingleton<AppImplementationFactory>();
+        return services;
+    }
+}
+
+internal sealed class AppImplementationFactory
+{
+    internal sealed record DIScopeHolder(IServiceProvider? ScopedServiceProvider);
+
+    private readonly DIScopeHolder _diScopeHolder;
+
+    public AppImplementationFactory(DIScopeHolder diScopeHolder)
+    {
+        _diScopeHolder = diScopeHolder;
+    }
+
+    private IServiceProvider _sp =>
+        // Right now we are just using the HttpContext to get the current scope,
+        // in the future we might not be always running in a web context,
+        // at that point we need to replace this
+        _diScopeHolder.ScopedServiceProvider
+        ?? throw new InvalidOperationException("Couldn't resolve DI container from current context");
+
+    public T GetRequired<T>()
+        where T : class => _sp.GetRequiredService<T>();
+
+    public T? Get<T>()
+        where T : class => _sp.GetService<T>();
+
+    public IEnumerable<T> GetAll<T>()
+        where T : class => _sp.GetServices<T>();
```

Comment:
I think it makes more sense to use the same names as IServiceProivder if we expose the same functionality.

1. This list of methods will work as a list of interfaces that we intend users to implement to override functionality.
2. More obvious to use the special method, instead of just "use this different service provider".
3. It documents and enforces that an interface should be retrieved as `IEnumerable<T>` instead of just fetching the first `T, so that we do that consistently.

Reply:
I like the idea of going factory.GetServiceIWant rather than factory.Get<ServiceIWant> because of the doc and autocomplete it gives

Reply:
> This list of methods will work as a list of interfaces that we intend users to implement to override functionality.

We can get the same effect from searching for attribute use though
![image](https://github.com/user-attachments/assets/a3024951-edbd-4251-906b-31f141fd865b)

> More obvious to use the special method, instead of just "use this different service provider".

In this case there are build errors if one tries to inject `IEFormidlingMetadata` using constructors or `IServiceProvider` directly though (and so you _have_ to use the the different service provider/factory regardless of how we implement it). In any case, if we are developing in the context of some feature, these methods living somewhere else probably won't matter in terms of "obviousness"? In that scenario you are thinking about how to implement the feature, right?

> It documents and enforces that an interface should be retrieved as IEnumerable<T> instead of just fetching the first `T, so that we do that consistently.

Isn't that obvious depending on the `T`? Presumably the person developing the code requiring the `T` also knows how many `T`s there are/could be?

> I like the idea of going factory.GetServiceIWant rather than factory.Get<ServiceIWant> because of the doc and autocomplete it gives.

Again, at the time one is reaching for `ServiceIWant`, you already know what you need right? I'm struggling a bit with these scenarios that are raised 😄

Reply:
I'm fine with both ways, with a slight preference for what @martinothamar already did, but with the same method names as IServiceProvider.